File: ios/Classes/EncryptionStrategy.swift
import Foundation
import LocalAuthentication

// Abstraction/Protocol class of EncryptionStrategy
// @available(iOS 11.3, *)
protocol EncryptionStrategy {
    // create and store private key to secure enclave
    func generateKeyPair(accessControlParam: AccessControlParam) throws -> SecKey
    
    // remove key from secure enclave
    func removeKey(tag: String) throws -> Bool
    
    // get status SecKey key from secure enclave (private method)
    func isKeyCreated(tag: String) throws -> Bool?
    
    // get publicKey key from secure enclave
    func getPublicKey(tag: String) throws -> String?
    
    // encryption
    func encrypt(message: String, tag: String) throws -> FlutterStandardTypedData?
    
    // decryption
    func decrypt(message: Data, tag: String) throws -> String?
    
    // sign
    func sign(tag: String, message: Data) throws -> String?
    
    // verify
    func verify(tag: String, plainText: String, signature: String) throws -> Bool
}

File: android/src/main/kotlin/id/my/anggaaryas/flutter_secure_enclave/EncryptionStrategy.kt
package id.my.anggaaryas.flutter_secure_enclave

import java.security.KeyPair
import id.my.anggaaryas.flutter_secure_enclave.model.*

interface EncryptionStrategy {
    @Throws(Exception::class)
    fun generateKeyPair(accessControlParam: AccessControlParam): KeyPair

    @Throws(Exception::class)
    fun removeKey(tag: String): Boolean

    @Throws(Exception::class)
    fun isKeyCreated(tag: String): Boolean?

    @Throws(Exception::class)
    fun getPublicKey(tag: String): String?

    @Throws(Exception::class)
    fun encrypt(message: String, tag: String): ByteArray?

    @Throws(Exception::class)
    fun decrypt(message: ByteArray, tag: String): String?

    @Throws(Exception::class)
    fun sign(tag: String, message: ByteArray): String?

    @Throws(Exception::class)
    fun verify(tag: String, plainText: String, signature: String): Boolean
}